### PR TITLE
[SPARK-33857][SQL] Unify the default seed of random functions

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Unify the seed of random functions
1. Add a hold place expression `UnresolvedSeed ` as the defualt seed.
2. Change `Rand`,`Randn`,`Uuid`,`Shuffle` default seed to `UnresolvedSeed `.
3. Replace `UnresolvedSeed ` to real seed at `ResolveRandomSeed` rule.

### Why are the changes needed?

`Uuid` and `Shuffle` use the `ResolveRandomSeed` rule to set the seed if user doesn't give a seed value. `Rand` and `Randn` do this at constructing.

It's better to unify the default seed at Analyzer side since we have used `ExpressionWithRandomSeed` at streaming query.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass exists test and add test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by maropu** [comment]
Just a question; any system having uuid with a seed param? I couldn't find it though.

**Comment by ulysses-you** [comment]
@maropu thanks for the thought.
> any system having uuid with a seed param?

I haven't seen it too, but before this PR Spark has already support the seed param.

I'd say the main change of this PR is to keep the seed initialize only once, in case of copy. For this idea, we make all the rand functions (include uuid) have the same seed param.

**Comment by maropu** [comment]
Ah, I see. If so, I think its better to focus on the refactoring for the seed initialization and not to support the new feature: `select uuid(1)` in this PR.

**Review by dongjoon-hyun — COMMENTED** [review]
This seems too intrusive for the goal; `select uuid(1)`. Could you simplify this PR?

**Comment by dongjoon-hyun** [comment]
Oh, I commented before reading @maropu 's last comment. Ya. This looks like two orthogonal purposes.

**Comment by ulysses-you** [comment]
thanks @maropu @dongjoon-hyun , will to narrow the goal and make this PR focus on one thing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala [hunk 1]
```diff
@@ -62,9 +64,19 @@ abstract class RDG extends UnaryExpression with ExpectsInputTypes with Stateful
...
  */
 trait ExpressionWithRandomSeed {
+  def seedExpression: Expression
   def withNewSeed(seed: Long): Expression
 }
 
+/**
+ * A place holder expression used in random functions, will be replaced after analyze.
+ */
+case class DefaultSeed() extends Expression with Unevaluable {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:74]: `case object UnresolvedSeed extends Expression with Unevaluable`?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala]: yes, it's better.

**Comment by ulysses-you** [comment]
I think it's ok to review this. cc @maropu @dongjoon-hyun @cloud-fan

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala [hunk 2]
```diff
@@ -83,14 +87,13 @@ trait ExpressionWithRandomSeed {
...
-  extends RDG with ExpressionWithRandomSeed {
+case class Rand(child: Expression, hideSeed: Boolean = false) extends RDG {
 
-  def this() = this(Literal(Utils.random.nextLong(), LongType), true)
+  def this() = this(UnresolvedSeed, true)
 
   def this(child: Expression) = this(child, false)
 
-  override def withNewSeed(seed: Long): Rand = Rand(Literal(seed, LongType))
+  override def withNewSeed(seed: Long): Rand = Rand(Literal(seed, LongType), hideSeed)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:96]: Ur, is this related to this PR, @ulysses-you ? This looks like an orthogonal bug fix a little to me. I'm wondering if we we have a test coverage for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala [hunk 3]
```diff
@@ -134,14 +137,13 @@ object Rand {
...
-  extends RDG with ExpressionWithRandomSeed {
+case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
 
-  def this() = this(Literal(Utils.random.nextLong(), LongType), true)
+  def this() = this(UnresolvedSeed, true)
 
   def this(child: Expression) = this(child, false)
 
-  override def withNewSeed(seed: Long): Randn = Randn(Literal(seed, LongType))
+  override def withNewSeed(seed: Long): Randn = Randn(Literal(seed, LongType), hideSeed)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:146]: ditto.

**Review by dongjoon-hyun — COMMENTED** [review]
In general, this PR looks good, but I have a question on the change because SPARK-33857 is an improvement PR which lands only at `master` branch.

If we are correcting some fix in this PR, we should split this PR by filing a new JIRA for that and backport to the release branches.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:96]: @dongjoon-hyun thanks for the careful check. Yes, it could be thought as bug fix. See the [jenkins test result](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133209/testReport/org.apache.spark.sql/SQLQuerySuite/SPARK_31594__Do_not_display_the_seed_of_rand_randn_with_no_argument_in_output_schema/) before this commit [a3c9f0b](https://github.com/apache/spark/pull/30864/commits/a3c9f0b1410f0b3b367c70dc534b1c0d0b08e137).

And I think for early branch, we just need to backport this place and leave this PR for master. But also, I'm OK to send a PR to fix this first, then go back to this PR.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:96]: This is an orthogonal bug fix, but it only affects the `sql` method. I don't have a strong opinion about backporting the fix or not.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala:96]: Got it. Thanks, @cloud-fan .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master for Apache Spark 3.2.0.
Thank you, @ulysses-you and all.

Merry Christmas and Happy New Year!

**Comment by maropu** [comment]
late lgtm.

**Comment by ulysses-you** [comment]
thank you all ! Merry Christmas !
